[tutorial] 데이터 핸들링 - Core, ORM으로 행 조회하기를 작성한다 (#18)
SoogoonSoogoonPythonists/sqlalchemy-for-pythonist: tutorial/5. 데이터 핸들링하기 - Select구문.md

# 5. 데이터 select하기 

<br>

> Core와 ORM을 이용해 데이터 Select하기 - 이번 섹션은 SQLAlchemy에서 가장 자주 쓰이느 Select 생성자에 대해서 다룹니다. Select 생성자는 Core 및 ORM을 이용한 어플리케이션 구문을 반환합니다.

Reviewer: ```suggestion
Core와 ORM을 이용해 데이터 Select하기 - 이번 섹션은 SQLAlchemy에서 가장 자주 쓰이는 Select 생성자에 대해서 다룹니다. Select 생성자는 Core 및 ORM을 이용한 어플리케이션 구문을 반환합니다.
```
쓰이는이 맞는 맞춤법이라고 생각해서 리뷰했습니다. ^^

> # 5. 데이터 select하기 

Reviewer 2: ```suggestion
# 5. 데이터 핸들링 - Core, ORM으로 행 조회하기
```

좀 더 목적지향적, 직관적인 문체로 바꿔 보았습니다.
파일 이름도 이걸로 바꿔주시면 좋을거 같아요.

<br>

> Core와 ORM을 이용해 데이터 Select하기 - 이번 섹션은 SQLAlchemy에서 가장 자주 쓰이는 Select 생성자에 대해서 다룹니다. Select 생성자는 Core 및� ORM을 이용한 어플리케이션 구문을 반환합니다.

Reviewer 2: ```suggestion
이번 섹션에서는 SQLAlchemy에서 가장 자주 쓰이는 Select에 대해서 다룹니다.
```

간결 + 특수 문자 제거했습니다.

Author: 아 ㅋㅋㅋㅋ 훨씬 간결하네요 감사합니다.

<br>

## select() SQL 구문 생성자
select() 생성자는 insert() 생성자와 같은 방식으로 쿼리문을 만들 수 있습니다.

```python
>>> from sqlalchemy import select
>>> stmt = select(user_table).where

> # name 컬럼에 접근하고 싶다면 테이블명.c.name
> (user_table.c.name == 'spongebob')

Reviewer 2: ```suggestion
(user_table.c.name == 'spongebob')  # name 컬럼에 접근하고 싶다면 테이블명.c.name
```

주석이 코드 옆으로 가는게 가독성이 좋아보여용

> >>> print(stmt)

Reviewer 2: ```suggestion
>>> print(stmt)
"""
SELECT user_account.id, user_account.name, user_account.fullname
FROM user_account
WHERE user_account.name = :name_1
"""
```

```

마찬가지로 쿼리문을 실행시키 위해 같은 레벨의 SQL 생성자들(select, insert, update, create등)처럼 execution 메서드에 쿼리를 넣어 실행시킬 수 있습니다.

```python
>>> with engine.connect() as conn:
...     for row in conn.execute(stmt):
...         print(row)
(1, 'spongebob', 'Spongebob Squarepants')
```

한편 ORM을 이용해 select() 생성자를 사용할 때는 <code>Session.exeucte()</code>를 사용해야합니다.
결과는 마찬가지로 <code>Row</code>객체를 반환하지만

> 완전한 엔티티를 포함하는 결과를 반환합니다.

Reviewer 2: > 완전한 엔티티를 포함하는 결과를 반환합니다.

완전한 엔티티가 뭔가요?? 위에랑 구체적으로 어떤 차이가 있는거에요? 이를 좀 더 적어주시면 좋을거 같아요!

> 한편 ORM을 이용해 select() 생성자를 사용할 때는 <code>Session.exeucte()</code>를 사용해야합니다.
> 결과는 마찬가지로 <code>Row</code>객체를 반환하지만
> 완전한 엔티티를 포함하는 결과를 반환합니다.

Reviewer 3: ```suggestion
한편 ORM을 이용해 select() 생성자를 사용할 때는 <code>Session.exeucte()</code>를 사용해야합니다.</br>
결과는 마찬가지로 <code>Row</code>객체를 반환하지만 완전한 엔티티를 포함하는 결과를 반환합니다.
```
가독성좋게 사용해보았습니다!

```python
>>> stmt = select(User).where(User.name == 'spongebob')
# User 클래스의 인스턴스 안에 있는 각 row 들을 출력
>>> with Session(engine) as session:
...     for row in session.execute(stmt):
...         print(row)

```

<br>

##  FROM절 세팅하기

select로 선택하고 싶은 컬럼들 혹은 테이블을 
select() 함수의 인자로도 사용할 수 있습니다.

각 컬럼들을 Core를 이용해 조회하고자

> <code>Table.c</code> 접근자(accessor)에서 <code>column</code>객체에 접근합니다.

Reviewer 2: 이게 무슨 말인지 잘 모르겠어요 ㅜ
말로 하는거 보다 그냥 코드 예시로 한방에 보여주면 좋을거 같은데...

여기서 한번에 설명할 수 없다면 이 구문은 차라리 생략해버리는 것도 좋을거 같습니다.

Author: 이 부분은 삭제하겠습니다

<br>

> ### ORM 엔티티와 and 컬럼 조회하기

Reviewer 2: ```suggestion
### ORM 엔터티 및 열 조회
```

>  ORM <code>Session.execute()</code>메서드를 이용해 쿼리문을 실행 시킬 때 <code>User</code> 클래스와 같은 전체 엔티티를 사용하면 결과 세트에서 

Reviewer 2: ```suggestion
 ORM `Session.execute()` 메서드를 이용해 쿼리문을 실행 시킬 때 `User` 클래스와 같은 전체 엔티티를 사용하면 결과 세트에서 
```

마크다운으로 수정했습니다.

 각 row에 엔티티가 포함되어 리턴이 됩니다.

```python

> >>> row = session.execute(select(User)).first

Reviewer 2: ```suggestion
>>> row = session.execute(select(User)).first()
```

>>> row[0]
User(id=1, name='spongebob', fullname='Spongebob Squarepants')
```

> 또는 클래스 바인딩 어트리뷰트를 사용하여 결과물 <code>row</code>내의 ORM 엔티티 컬럼을 select 할 수 있습니다.

Reviewer 2: ```suggestion
또는 엔티티 속성을 이용하여 `row` 내의 ORM 엔티티 컬럼을 가져올 수 있습니다.
```

```python
>>> prow = session.execute(select(User.name, User.fullname)).first()
>>> row
('spongebob', 'Spongebob Squarepants')
```

또는 믹스인 방식으로, 행의 첫 번째 요소에는 User entity의 이름 속성을 선택하고 두 번째 요소의 전체 Address entity를 넣어 다음과 같은 접근 방식도 혼용할 수 있습니다.

```python
>>> session.execute(
...     select(User.name, Address).
...     where(User.id==Address.user_id).
...     order_by(Address.id)
... ).all()

[('spongebob', Address(id=1, email_address='spongebob@sqlalchemy.org')),
('sandy', Address(id=2, email_address='sandy@sqlalchemy.org')),
('sandy', Address(id=3, email_address='sandy@squirrelpower.org'))]
```

<br>

### 라벨링된 데이터 조회하기

<code>ColumnElement.label() </code>는  SQL 
쿼리 실행 결과물에서 컬럼에 특정한 이름(라벨)을 지을 수 있게 도와줍니다. 이러한 라벨은 결과물 세트에서 각 row에서 참조할 수도 있습니다.

```python
>>> from sqlalchemy import func, cast
>>> stmt = (
...     select(

> ...         ("Username: " + user_table.c.name).label("username"),

Reviewer 2: `("Username: " + user_table.c.name)` 

이렇게 문자열과 `ColumnElement` 를 `+` 연산하면 `ColumnElement` 가 나오나요??

Author: ## 문자열 +  `Column` 하면 `BinaryExpression`객체가 나옵니다.
` user_table.c.name`은 `<class 'sqlalchemy.sql.schema.Column'>`인데
`"Username: " + user_table.c.name`하면 `<class 'sqlalchemy.sql.elements.BinaryExpression'>`되네요!!!
`+`하니까 타입이 `BinaryExpression`로 바껴요.

### `BinaryExpression`에 관한 공식 문서 
[BinaryExpression](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=binaryexpression#sqlalchemy.sql.expression.BinaryExpression)
A `BinaryExpression` is generated automatically whenever two column expressions are used in a Python binary expression:

근데  string + column 해도 BinaryExpression 나옵니다 ㅎㅎ

```python 
>>> from sqlalchemy.sql import column
>>> column('a') + column('b')
<sqlalchemy.sql.expression.BinaryExpression object at 0x101029dd0>
>>> print(column('a') + column('b'))
a + b
```

### 튜토리얼에 나온 예제 
```python
from sqlalchemy import select, MetaData, Table, Column, Integer, String, ForeignKey
from sqlalchemy.orm import registry, relationship

metadata = MetaData()

user_table = Table(
    'user_account',
    metadata,
    Column('id', Integer, primary_key=True),
    Column('name', String(30)),
    Column('fullname', String(100)) #String으로 안보내지고 String(100) 으로 지정    
)

address_table = Table(
    'address',
    metadata,
    Column('id', Integer, primary_key=True),
    Column('user_id', ForeignKey('user_account.id'), nullable=False),
    Column('email_address', String(100), nullable=False)    #String으로 안보내지고 String(100) 으로 지정    

)
print(type(user_table.c.name)) # <class 'sqlalchemy.sql.schema.Column'>
print(type(type("Username: " + user_table.c.name)) # <class 'sqlalchemy.sql.elements.BinaryExpression'>
```

...     ).order_by(user_table.c.name)
... )
>>> with engine.connect() as conn:
...     for row in conn.execute(stmt):
...         print(f"{row.username}")
Username: patrick
Username: sandy
Username: spongebob
```

<br>

### 스트링 데이터 조회하기

보통 <code>Select</code>객체나 <code>select()</code> 생성자를 이용 할 때는 컬럼을 조회하는 경우가 다수지만 가끔은 임의로 스트링 문자열과 함께 컬럼을 조회를 해야하는 경우가 있습니다. 이번 섹션에서는 스트링 데이터를 조회 하는 방법에 다룹니다.

> 아래는 text() 생성자를 이용해 `some label`과 함께 특정 컬럼을 같이 출력하는 예제입니다.

Reviewer 2: ```suggestion
아래는 `text()` 생성자를 이용해 `some label`과 함께 특정 컬럼을 같이 출력하는 예제입니다.
```

코드 구문은 인라인  코드블락으로 만들어주시면 가독성 좋습니다.  
다른 부분들도 마찬가지입니당 ㅎㅎ

```python
>>> from sqlalchemy import text
>>> stmt = (
...     select(
...         text("'some phrase'"), user_table.c.name
...     ).order_by(user_table.c.name)
... )
>>> with engine.connect() as conn:
...     print(conn.execute(stmt).all())

> ('some phrase', 'patrick'), ('some phrase', 'sandy'), ('some phrase', 'spongebob')]

Reviewer 2: ```suggestion
[('some phrase', 'patrick'), ('some phrase', 'sandy'), ('some phrase', 'spongebob')]
```

```

> 하지만 일반적으로 literal_column() 생성자를 이용하면 스트링 문자열을 출력하는데 더 편리한 기능을 제공해줍니다.

Reviewer 2: ```suggestion
하지만 일반적으로 `literal_column()` 생성자를 이용하면 스트링 문자열을 출력하는데 더 편리한 기능을 제공해줍니다.
```

> 이 생성자는 text() 와 매우 비슷하지만, 

Reviewer 2: ```suggestion
이 생성자는 `text()` 와 매우 비슷하지만, 
```

> 임의로 SQL 형식을 출력하기보다 보다 명시적으로 하나의

Reviewer 3: ```suggestion
임의로 SQL 형식을 출력하기보단 보다 명시적으로 하나의
```

Reviewer 4: ```suggestion
임의로 SQL 형식을 출력하기보단 명시적으로 하나의
```

> <code>column</code>을 표시하고 서브쿼리를 이용 할 수 있게 라벨 기능을을 제공합니다.

Reviewer 2: ```suggestion
`column`을 표시하고 서브쿼리를 이용 할 수 있게 라벨 기능을 제공합니다.
```

```python
>>> from sqlalchemy import literal_column
>>> stmt = (
...     select(
...         literal_column("'some phrase'").label("p"), user_table.c.name
...     ).order_by(user_table.c.name)
... )
>>> with engine.connect() as conn:
...     for row in conn.execute(stmt):
...         print(f"{row.p}, {row.name}")
some phrase, patrick
some phrase, sandy
some phrase, spongebob

```

> > 하지만 text() 또는 literal_column()을 사용한다는 것은  때는 리터럴 값이 아닌 SQL 구문을 사용한다는 뜻입니다. 따라서 우리는 렌더링하려는 SQL에 필요한 ""와 같은 쌍따옴표등을 추가해야합니다.

Reviewer 3: ```suggestion
> 하지만 text() 또는 literal_column()을 사용한다는 것은 리터럴 값이 아닌 SQL 구문을 사용한다는 뜻입니다. 따라서 우리는 렌더링하려는 SQL에 필요한 ""와 같은 쌍따옴표등을 추가해야합니다.
```

<br>

## WHERE절

SQLAlchemy를 사용하면 Python 연산자를 사용하여 이름 = 'thead' 또는 user_id > 10과 같은 SQL 식을 작성할 수 있습니다.

```python
>>> print(user_table.c.name == 'squidward')

> user_account.name = :name_1

Reviewer 2: 오.. `==` 연산자가 참/거짓을 판단하는 오퍼레이터로 안쓰이고 이렇게 표현되는게 신기하네요 ...

>>> print(address_table.c.user_id > 10)
address.user_id > :user_id_1

> 

Reviewer 2: 여기에 ``` 가 빠진듯해요

where 절을 통한 join을 구현 할 경우 아래와 같이 작성 가능합니다.
```python
>>> select(address_table.c.email_address)
...     where(user_table.c.name == 'squidward').
...     where(address_table.c.user_id == user_table.c.id)
... 

>>> # 위와 같은 표현
...     select(address_table.c.email_address).
...     where(
...          user_table.c.name == 'squidward',
...          address_table.c.user_id == user_table.c.id
...     )
...
```

and_(), or_()등의 연결 구문도 구현가능합니다.

```python
>>> from sqlalchemy import and_, or_
>>>     select(Address.email_address).
...     where(
...         and_(
...             or_(User.name == 'squidward', User.name == 'sandy'),
...             Address.user_id == User.id
...         )
...     )
... 
```

단순히 같은지, 아닌지 비교하는 경우(equality) <code>Select.filter_by()</code>도 자주 사용됩니다.
```python
>>>select(User).filter_by(name='spongebob', fullname='Spongebob Squarepants')
```

<br>

## FROM절과 JOIN 명시하기

> FROM 절은 보통 Select의 선택된 컬럼들에 있는 테이블을 렌더합니다.(table_name.column)

Reviewer 2: 이 구문들이 무슨말인지 잘 모르겠어요.
우리가 무엇을 할 수 있는지만 말해주고, 코드만 보여주면 좋을거 같습니다.

> join을 이용해 두 개 이상의 테이블을 참조해 select 해야하는 경우가 있는데 이 상황에서 두 개의 메서드를 선택할 수 있다.

Reviewer 2: 갑자기 존대 -> 반말로 변하는 문장들이 있는데, 이런 문체 모두 찾아서 바꿔주시면 좋을듯 해용

첫 째는 <code>Select.join_from()</code>로 
JOIN구문에서 명시적으로 왼쪽테이블, 오른쪽 테이블을 지정할 수 있게 합니다.
두 번째는  <code>Select.join()</code>로 오직 오른쪽 테이블만 명시되어 있다.

>>> print(select(user_table.c.name, address_table.c.email_address)) # 두 개의 테이블 참조
```

```python
>>>     select(user_table.c.name, address_table.c.email_address).
...     join(address_table)
```

> 한편 우리가 FROM절에 추가 옵션을 주고 싶을 때는 Select.select_frim()메서드를 사용합니다.

Reviewer 2: ```suggestion
한편 우리가 FROM절에 추가 옵션을 주고 싶을 때는 `Select.select_from()` 메서드를 사용합니다.
```

```python
>>> print(
...     select(address_table.c.email_address).
...     select_from(user_table).join(address_table)
... )

> 
> 
> SELECT address.email_address
> FROM user_account JOIN address ON user_account.id = address.user_id

Reviewer 2: ```suggestion
"""
SELECT address.email_address
FROM user_account JOIN address ON user_account.id = address.user_id
"""
```

파이썬 코드블락에 들어가는 문자열은 이렇게 `"""` 로 묶어주시면 좋을거 같아요.

Reviewer 2: 이렇게 적용안되어 있는 다른 부분들도 수정해주시면 좋을거 같아요. 따로 코멘트 달지 않을게요 ㅎㅎ

만약에 Join의 대상인 왼쪽 테이블과 오른쪽 테이블이 이러한 제약 key가 없다거나 여러개의 key를 필요로 할 경우 ON 절을 직접 지정해야 합니다. 이러한 기능은 Select.join()과 Select 둘 다 제공하고 join_from()은 ON 절에 대한 추가 인수를 허용합니다.

```python
>>> print(
...     select(address_table.c.email_address).
...     select_from(user_table).
...     join(address_table, user_table.c.id == address_table.c.user_id)
... )

SELECT address.email_address
FROM user_account JOIN address ON user_account.id = address.user_id
```
<br>

### OUTER, FULL Join

<code>Select.join()</code>과 <code>Select.join_from()</code>메서드 둘 다 키워드 인자로 <code>Select.join.isouter</code>,
<code>Select.join.full </code>를 받고 이를 각 각
LEFT OUTER JOIN, FULL OUTER JOIN으로 렌더링 합니다.

```python
>>> print(
...     select(user_table).join(address_table, isouter=True)
... )

SELECT user_account.id, user_account.name, user_account.fullname
FROM user_account LEFT OUTER JOIN address ON user_account.id = address.user_id
```
<br>

## ORDER BY, GROUP BY, HAVING
  
<br>

### ORDER BY

<code>Select.order_by()</code>를 이용해 order_by()기능을 구현할 수 있고 오름차순, 내림차순은 

> <code>ColumnElement.asc()</code><code>ColumnElement.desc()</code>를 

Reviewer 4: ```suggestion
<code>ColumnElement.asc()</code>, <code>ColumnElement.desc()</code>를
```
붙어있으니 조금 보기 불편하네요. 271번째 라인처럼 콤마를 통해서 구분지어주면 좋을 것 같습니다.

<code>order_by()</code> 인자값 앞에서 사용하면 됩니다.

```python
>>> print(select(user_table).order_by(user_table.c.name))

SELECT user_account.id, user_account.name, user_account.fullname
FROM user_account ORDER BY user_account.name

>>> print(select(User).order_by(User.fullname.desc()))

SELECT user_account.id, user_account.name, user_account.fullname
FROM user_account ORDER BY user_account.fullname DESC
```
<br>

### GROUP BY, Having을 이용한 집계함수 

> SQL에서 집계 함수를 사용하면 여러 행에 걸쳐 있는 합쳐 단일 결과를 생성할 수 있습니다. 예를 들어 카운팅, 평균 계산은 물론 값 집합에서 최대값 또는 최소값을 찾는 작업을 할 수 있다. 

Reviewer 4: ```suggestion
SQL에서 집계 함수를 사용하면 여러 행에 걸쳐 있는 데이터를 합쳐 단일 결과를 생성할 수 있습니다. 예를 들어 카운팅, 평균 계산은 물론 최대값 또는 최소값을 찾는 작업을 할 수 있습니다. 
```
데이터를 합치는 것 맞죠??
최대 최소의 경우 값 집합에서 찾는 것이 당연한 것 같아 빼봤습니다.

 SQLAlchemy는 func라고 하는 네임스페이스를 사용하여 SQL 함수를 개방형 방식으로 제공합니다. 이것은 특정 SQL 함수에 전달할 0개 이상의 인수가 지정된 경우 함수의 새 인스턴스를 만드는 특수 생성자 개체입니다. 예를 들어 user_account.id 열에 대해 SQL COUNT() 함수를 렌더링하려면 count() 이름을 호출합니다.

```python
>>> from sqlalchemy import func
>>> count_fn = func.count(user_table.c.id)
```

SQLAlchemy는 Select.group_by() 및 Select.having() 메서드를 사용하여 이 group_by와 having 기능을 제공합니다. 아래에서는 둘 이상의 주소를 가진 사용자에 대한 사용자 이름 필드 및 주소 수를 선택합니다.

```python
>>> with engine.connect() as conn:
...     result = conn.execute(
...         select(User.name, func.count(Address.id).label("count")).
...         join(Address).
...         group_by(User.name).
...         having(func.count(Address.id) > 1)
...     )
...     print(result.all())

[('sandy', 2)]
```

<br>

### 라벨을 통해 Ordering과 Grouping하기 

> 라벨을 통해 ordering, grouping하려면 스트링 텍스트의 이름을 <code>Select.order_by()</code> 또는 <code>Select.group_by()</code> 메서드에 전달하여 사용할 수 있습니다. 전달된 텍스트는 곧바로 렌더링되지 않습니다. 대신 컬럼절에 전잘 된 이름과 컨텍스트에 전달 된 이름이 일치하지 않으면 오류가 발생합니다.

Reviewer 4: ```suggestion
라벨을 통해 ordering, grouping하려면 스트링 텍스트의 이름을 <code>Select.order_by()</code> 또는 <code>Select.group_by()</code> 메서드에 전달하여 사용할 수 있습니다. 전달된 텍스트는 곧바로 렌더링되지 않습니다. 대신 컬럼절에 전달 된 이름과 컨텍스트에 전달 된 이름이 일치하지 않으면 오류가 발생합니다.
```

```python
>>> from sqlalchemy import func, desc
>>> stmt = select(

> ...         Address.user_id,
> ...         func.count(Address.id).label('num_addresses')).\
> ...         group_by("user_id").order_by("user_id", desc("num_addresses"))

Reviewer 2: ```suggestion
...     Address.user_id,
...     func.count(Address.id).label('num_addresses')).\
...     group_by("user_id").order_by("user_id", desc("num_addresses"))
```

인덴트 맞춰주세용 ㅎㅎ

```
<br>

## 별칭 사용하기

<br>

### ORM 엔티티 별칭

 SQLAlchemy에서  as 구문을 통해 만들어진 이름들은  <code>FromClause</code> 객체를 사용하여 Core에 구성된 <code>FromClause.alias() method</code>
 를 이용하여 사용자에게 보여집니다.
  
SQLAlchemy에서는 별칭들은 <code>FromClause.Alias</code>생성자로 사용가능합니다.
<code>Alias</code>생성자는 Alias.c 컬렉션 내에 <code>Column</code>객체 네임스페이스도 있다는 점에서 <code>Table</code>생성자와 비슷합니다.

```python
>>> user_alias_1 = user_table.alias()

> >>> user_alias_2 = user_table.alias()

Reviewer 2: `alias()` 의 결과물이 무엇인지 읽는 사람 입장에서는 모를거 같아요 ㅎㅎ
이건 언제 쓰이는거고, 용도가 정확히 뭘까요??

```

<br>

## 서브쿼리와 CTE

> 이 섹션에서는 일반적으로 SELECT의 FROM 절에 있는 서브 쿼리에 대해 설명합니다. 서브쿼리와 유사한 방식으로 사용되지만 추가 기능이 포함된 CTE도 다룹니다.

Reviewer 2: 전 CTE에 대해서 처음 들어봤는데, 처음 알게된 사람을 위해서 참고 링크라도 남겨주시면 좋을거 같아요.

SQL에서 SELECT 문은 UNION 또는 UNION ALL을 수행하여 결과물을 합칠 수 있으며 하나 이상의 쿼리에서 생성된 모든 행들을 집합을 생성합니다. <code>INTERSECT [ALL], EXCREST [ALL]</code>도 같은 역할을 수행할 수 있다. 

SQLAlchemy의 <code>Select</code> 생성자는 <code>union(), crossect(), except_all(),union_all(), crossect_all() ,except_all()</code>를 지원합니다.

이러한 함수로 생성되는 함수는 <code>CompoundSelect</code>이며, 이는 메서드가 더 적다는 점을 제외하고 <code>Select</code>와 비슷한 동일한 방식으로 사용됩니다.
```python
>>> from sqlalchemy import union_all
>>> stmt1 = select(user_table).where(user_table.c.name == 'sandy')
>>> stmt2 = select(user_table).where(user_table.c.name == 'spongebob')
>>> u = union_all(stmt1, stmt2)
>>> with engine.connect() as conn:
...     result = conn.execute(u)
...     print(result.all())
```
<code>CompoundSelect</code>를 하위 쿼리로 사용하려면 <code>SelectBase.subquery()</code> 메서드를 사용하여 FromClause.c 컬렉션을 생성하는 Subquery() 메서드를 제공합니다.

```python
>>> u_subq = u.subquery()
>>> stmt = (
...     select(u_subq.c.name, address_table.c.email_address).
...     join_from(address_table, u_subq).
...     order_by(u_subq.c.name, address_table.c.email_address)
... )
>>> with engine.connect() as conn:
...     result = conn.execute(stmt)
...     print(result.all())
```

<br>

## EXISTS 서브쿼리들

> SQL EXISTS 키워드는  SELECT 문이 행을 반환하는지 여부에 따라 True 또는 false를 반환합니다.

Reviewer 4: ```suggestion
SQL EXISTS 키워드는  SELECT 문이 행을 반환하는지 여부에 따라 True 또는 False를 반환합니다.
```

 EXISTS 키워드 스칼라 서브 쿼리와 함께 사용되는 연산자입니다.

> SQLAlchemy는 ScalarSelect 객체의 종류 중 하나로 Exists 객체를 갖고 있는데 이는 EXISTS 서브쿼리를 생성하고 SelectBase.exists()라는 메서드를 통해 Exists객첼르 생성할 수 있습니다.

Reviewer 2: ```suggestion
`ScalarSelect` 객체의 종류 중 하나로 EXISTS 서브쿼리를 생성하는 `Exists` 객체가 있습니다. 이는 `SelectBase.exists()` 메서드를 통해 생성할 수 있습니다.
```

```python
>>> subq = (
...     select(func.count(address_table.c.id)).
...     where(user_table.c.id == address_table.c.user_id).
...     group_by(address_table.c.user_id).
...     having(func.count(address_table.c.id) > 1)
... ).exists()
>>> with engine.connect() as conn:
...     result = conn.execute(
...         select(user_table.c.name).where(subq)
...     )
...     print(result.all())

[('sandy',)]
```

한편 EXISTS 구문은 부정으로 사용되지 않는 경우가 더 많습니다.
```python
# 이메일 주소가 없는 유저네임을 선택하는 쿼리문입니다.
# "~" 연산이 들어간 부분을 확인해보세요
>>> subq = (
...     select(address_table.c.id).
...     where(user_table.c.id == address_table.c.user_id)
... ).exists()
>>> with engine.connect() as conn:
...     result = conn.execute(
...         select(user_table.c.name).where(~subq) 
...     )
...     print(result.all())
```

<br>

## SQL 함수 다뤄보기

> SQL 함수 작업

Reviewer 3: ```suggestion
### SQL 함수 작업
```

이 섹션 앞부분의 GROUP BY / HAVING을 사용한 집계 함수에서 처음 소개된 <code>func</code> 객체는
새로운 <code>Function</code> 객체를 생성하기 위한 팩토리 역할을 합니다.  
여기서 <code>Function</code> 객체는 <code>select()</code>와 같은 구문을 사용 할 때   
이름,괄호(항상 그렇지는 않지만) 종종 인자값등 으로 구성된 SQL 함수를 생성합니다.  
일반적인 SQL 함수는 다음과 같습니다.

> - <code>count()</code> : 집계 함수로 몇개의 행이 반환되었지 세서 그 수를 반환해줍니다.

Reviewer 3: ```suggestion
- <code>count()</code> : 집계 함수로 반환된 행의 수를 반환해줍니다.
```

  ```python
  >>> print(select(func.count()).select_from(user_table))
  
  SELECT count(*) AS count_1FROM user_account
  ```

> - <code>lower()</code> : 문자열 함수로 스트링 문자열을 소문자로 바꿔줍니다.

Reviewer 3: ```suggestion
- <code>lower()</code> : 문자열 함수로 문자열을 소문자로 바꿔줍니다. 
또는 대문자 문자열을 소문자로 바꿔주기도 합니다.
```

  ```python
  >>> print(select(func.lower("A String With Much UPPERCASE")))
  
  SELECT lower(:lower_2) AS lower_1
  ```
- <code>now()</code> : 현재 시간과 날짜를 반환해주는 함수입니다. 
  이 함수는 굉장히 흔하게 사용되는 함수이기에 SQLAlchemy는 서로 다른 백엔드에서 손쉽게 렌더링 할 수 있도록 도와줍니다. 
  ```` python
  #SQLite에서 CURRENT_TIMESTAMP함수로 쓰일 경우 
  >>> stmt = select(func.now())
  >>> with engine.connect() as conn:
  ...     result = conn.execute(stmt)
  ...     print(result.all())
  ````

<br> 

> ### 리턴 타입이 있는 함수Functions Have Return Types

Reviewer 2: 이런거... 정리가 되다 안된 느낌인데, 전반적으로 이런 부분들 다시 한번 검토해주시면 좋을거 같아요.

SQL에서도 데이터 타입이 있는데 SQLAlchemy에서는 "SQL 리턴 타입"이라고 말합니다.

아래와 <code>Function.type</code>속성을 이용해 SQL 함수의 SQL 리턴타입을 확인 할 수 있습니다.
```python
>>> func.now().type
DateTime()
```

이러한 SQL 리턴 유형은 길이가 긴 쿼리에서 함수 표현식을 사용할 때 중요합니다.
예를 들면 수학 연산자는 표현식의 데이터 유형이 <code>Integer</code> 또는 <code>Numeric</code> 
일때 더 잘 작동되고 JSON 접근자를 처리하려면 <code>JSON</code> 타입을 사용해야합니다.

> 한편 특정 클래스의 함수는 특정 열을 참조해야 하는 경우 컬럼 값 대신 전체 행을 반환하는데 이러한 함수를

Reviewer 4: ```suggestion
- 한편 특정 클래스의 함수는 특정 열을 참조해야 하는 경우, 컬럼 값 대신 전체 행을 반환하기도 합니다.  이러한 함수를 <Code>table valued functions. </code> 이라고 합니다.
```

```
<br> 

### WITHIN GROUP, FILTER등 특수한 지정자

"WITHIN GORUP"이라는 SQL 구문은 순서 집합 또는 가상 집합 그리고 집계함수와 함께 쓰입니다.
일반적인 순서 집합 함수는 <code>percentile_cont()(</code> 그리고 <code>rank()</code>를 포함하고 있습니다.
SQLAlchemy에서는 <code>rank, dense_rank, percentile_count, percentile_disc</code>가 구현되어 있고
각 각은 <code>FunctionElement.within_group()</code>메서드를 갖고 있습니다.

```python
>>> print(
...     func.unnest(
...         func.percentile_disc([0.25,0.5,0.75,1]).within_group(user_table.c.name)
...     )
... )

unnest(percentile_disc(:percentile_disc_1) WITHIN GROUP (ORDER BY user_account.name))
```
어떤 백엔드에서는 "FILTER"를 지원하는데 이는 <code> FunctionElement.filter()</code>메서드로 
사용이 가능합니다. 
```python
>>> stmt = select(
...     func.count(address_table.c.email_address).filter(user_table.c.name == 'sandy'),
...     func.count(address_table.c.email_address).filter(user_table.c.name == 'spongebob')
... ).select_from(user_table).join(address_table)
>>> with engine.connect() as conn:  
...     result = conn.execute(stmt)
...     print(result.all())

[...] ('sandy', 'spongebob')
[(2, 1)]
```

<br> 

### Table-Valued Functions
> 원문의 Table-Valued Functions 부분은 제가 이해하지 못했습니다. 
> 이에 대해 이해하신 분 있으시다면 이 부분에 기여 부탁드립니다. 감사합니다.

> 

Reviewer 2: ```suggestion
<br>
```

```
그래서 보통은 `text()`보다 `literal_column()`을 사용해 작은 따옴표가 결과물에 붙어져서 나오는 문제를 해결할 수 있습니다.  
여기에서  `text`와 `literal_column()`은 거의 비슷하지만  `literal_column()`은 명시적으로 컬럼을 의미하고, 서브쿼리나 다른 SQL 표현식에서 쓰일 수 있게 라벨링도 할 수 있습니다.  

```python
>>> from sqlalchemy import literal_column
>>> stmt = (
...     select(
...         literal_column("'some phrase'").label("p"), user_table.c.name
...     ).order_by(user_table.c.name)
... )
>>> with engine.connect() as conn:
...     for row in conn.execute(stmt):
...         print(f"{row.p}, {row.name}")
some phrase, patrick
some phrase, sandy
some phrase, spongebob

```

<br>

## WHERE절

SQLAlchemy를 사용하면 Python 연산자를 사용하여 `name` = 'thead' 또는 `user_id` > 10인 데이터만 출력하는 쿼리를 쉽게 작성할 수 있습니다.

```python
>>> print(user_table.c.name == 'squidward')
user_account.name = :name_1

>>> print(address_table.c.user_id > 10)
address.user_id > :user_id_1
```

`WHERE`절을 만들기 위해 `Select.where()`메서드안에 인자값을 넣어 사용할 수도 있습니다.

```python
>>> print(select(user_table).where(user_table.c.name == 'squidward'))
"""
SELECT user_account.id, user_account.name, user_account.fullname
FROM user_account
WHERE user_account.name = :name_1
"""
```

> where 절을 통한 join을 구현 할 경우 아래와 같이 작성 가능합니다.

Reviewer 2: ```suggestion
WHERE 절을 통한 JOIN을 구현해야 할 때, 아래와 같이 작성 가능합니다.
```

```python
>>> select(address_table.c.email_address).
...     where(user_table.c.name == 'squidward').
...     where(address_table.c.user_id == user_table.c.id)
... 

>>> # 위와 같은 표현
...     select(address_table.c.email_address).
...     where(
...          user_table.c.name == 'squidward',
...          address_table.c.user_id == user_table.c.id
...     )
...
```

`and_()`, `or_()`등의 연결 구문도 구현가능합니다.

```python
>>> from sqlalchemy import and_, or_

> >>>     select(Address.email_address).
> ...     where(
> ...         and_(
> ...             or_(User.name == 'squidward', User.name == 'sandy'),
> ...             Address.user_id == User.id
> ...         )
> ...     )
> ... 

Reviewer 2: ```suggestion
>>> print(
...     select(Address.email_address).
...         where(
...             and_(
...                 or_(User.name == 'squidward', User.name == 'sandy'),
...                 Address.user_id == User.id
...             )
...         )
...  )
"""
SELECT address.email_address
FROM address, user_account
WHERE (user_account.name = :name_1 OR user_account.name = :name_2)
AND address.user_id = user_account.id
"""
```

```

단순히 같은지, 아닌지 비교하는 경우(equality) `Select.filter_by()`도 자주 사용됩니다.
```python

> >>>select(User).filter_by(name='spongebob', fullname='Spongebob Squarepants')

Reviewer 2: ```suggestion
>>> print(
...     select(User).filter_by(name='spongebob', fullname='Spongebob Squarepants')
... )
"""
SELECT user_account.id, user_account.name, user_account.fullname
FROM user_account
WHERE user_account.name = :name_1 AND user_account.fullname = :fullname_1
"""
```

```

<br>

## FROM절과 JOIN 명시하기

앞에서 언급했지만 `FROM`절은 따로 명시하지 않아도 `select()`메서드의 인자에 넣은 컬럼들로 자동 세팅이 됩니다.
```python
# 따로 FROM 절을 명시하지 않았지만 FROM 절이 세팅되어 출력됩니다.
>>> print(select(user_table.c.name))
"""
SELECT user_account.name
FROM user_account
"""
```

만약 ` select()`의 위치 인자로  서로 다른 두 개의 테이블을  참조하는 컬럼을`,`(컴마)로 구분지어 넣을 수도 있습니다
```python
>>> print(select(user_table.c.name, address_table.c.email_address))
"""
SELECT user_account.name, address.email_address
FROM user_account, address
"""
```

서로 다른 두 개의 테이블을 `JOIN`조인하고 싶다면, 이용할 수 있는 메서드가 두 가지가 있는데요,

> 하나는 `Select.join()`메서드로 명시적으로 `JOIN` 할 왼쪽에 들어갈 테이블, 오른 쪽에 들어 갈 테이블을 직접 지정할 수 있습니다

Reviewer 2: ```suggestion
하나는 `Select.join()`메서드로 명시적으로 `JOIN` 할 왼쪽에 들어갈 테이블, 오른쪽에 들어 갈 테이블을 직접 지정할 수 있습니다
```

```python
>>> print(
...     select(user_table.c.name, address_table.c.email_address).
...     join_from(user_table, address_table)
... )
"""
SELECT user_account.name, address.email_address
FROM user_account JOIN address ON user_account.id = address.user_id
"""
```

> 다른 하나는 `select.join()`메서드로 오른쪽에 들어갈 테이블만 명시적으로 적어주고 

Reviewer 2: ```suggestion
다른 하나는 `Select.join()`메서드로 오른쪽에 들어갈 테이블만 명시적으로 적어주고 
```

나머지 테이블은 암시적으로 컬럼을 선택할 때 참조하게 합니다.
```python

> side is inferred:
> >>> print(
> ...     select(user_table.c.name, address_table.c.email_address).
> ...     join(address_table)
> ... )

Reviewer 2: ```suggestion
# 위와 동일한 표현이지만, 조인할 왼쪽 테이블(user_table)은 암시적으로 표현됩니다.
>>> print(
...     select(user_table.c.name, address_table.c.email_address).
...     join(address_table)
... )
```

"""
SELECT user_account.name, address.email_address
FROM user_account JOIN address ON user_account.id = address.user_id
"""
```
또는 `JOIN` 하는 두 개의 테이블을 조금 더 명시적으로 작성하고 싶거나 `FROM`절에 명시적인 추가 옵션을 주고 싶다면
 아래와 같이 작성 할 수도 있습니다.
```python
>>> print(
...     select(address_table.c.email_address).
...     select_from(user_table).join(address_table)
... )
"""
SELECT address.email_address
FROM user_account JOIN address ON user_account.id = address.user_id
"""
```

> `Select.select_from()`을 사용하는 또 다른 경우는 우리가 조회하고 싶은 컬럼들을 통해 암시적으로 FROM절을 세팅할 수 없는 경우입니다.   예를 들면 일반적인 SQL 쿼리문에서 `count(*)`를 조회하기 위해선  SQLAlchemy의 `sqlalchemy.sql.expression.func`를 사용해야합니다.

Reviewer 2: ```suggestion
`Select.select_from()`을 사용하는 또 다른 경우는 우리가 조회하고 싶은 컬럼들을 통해 암시적으로 `FROM` 절을 세팅할 수 없는 경우입니다.   예를 들면 일반적인 SQL 쿼리문에서 `count(*)`를 조회하기 위해선  SQLAlchemy의 `sqlalchemy.sql.expression.func`를 사용해야합니다.
```

```python
>>> from sqlalchemy import func
>>> print(select(func.count('*')).select_from(user_table))
"""
SELECT count(:count_2) AS count_1
FROM user_account
"""
```

<br>

### On절 세팅하기

그런데 뭔가 이상한게 있었죠? 

> 사실 이 전의 예제에서 `Select.select_from()`이나 ` select.join()`을 이용해 두 테이블을 `JOIN`할 때 암시적으로 `ON절`이 세팅되었답니다.    

Reviewer 2: ```suggestion
사실 이 전의 예제에서 `Select.select_from()`이나 `select.join()`을 이용해 두 테이블을 `JOIN`할 때 암시적으로 `ON`절이 세팅되었답니다.    
```

왜 자동으로  `ON` 절이 세팅 됐냐면, `user_table`객체와, `address_table` 객체가 ` ForeignKeyConstraint`, 즉 외부키 제약을 갖고 있어서 자동으로 세팅이 된 것입니다.  

> 만약에 `Join`의 대상인 두 개의 테이블에서 이러한 제약 key가 없을 경우 `ON` 절을 직접 지정해야 합니다. 이러한 기능은 `Select.join()`과 `Select.join_from()`메서드 둘 다 명시적으로 `ON`절을 세팅할 수 있습니다.

Reviewer 2: ```suggestion
만약에 `Join`의 대상인 두 개의 테이블에서 이러한 제약 key가 없을 경우 `ON`절을 직접 지정해야 합니다. 이러한 기능은 `Select.join()`나 `Select.join_from()`메서드에 파라미터 전달을 통해 명시적으로 `ON`절을 세팅할 수 있습니다.
```

"""
```
<br>

### 그룹함수 : GROUP BY, Having

SQL에서는 집계함수를 이용해 조회 된 여러개의 행들을 하나의 행으로 합칠 수도 있습니다.
집계함수의 예로 `COUNT()`, `SUM()`, `AVG()`등이 있습니다.

 SQLAlchemy에서는  `func`라는 네임스페이스를 이용해 SQL함수를 제공하는데, 이 `func`는
SQL함수의 이름이 주어지면  `Function`인스턴스를 생성합니다.

아래의 예제에서는 `user_account.id`컬럼을 SQL `COUNT()`함수에 렌더링 하기 위해 `count()`함수를 호출합니다.
```python
>>> from sqlalchemy import func
>>> count_fn = func.count(user_table.c.id)
>>> print(count_fn)
"""
count(user_account.id)
"""
```

SQL 함수에 관해서는 [SQL 함수 다루기]()에서 더 자세히 설명되어 있습니다.

다시 설명하자면
- `GROUP BY`는 조회된 행들을 특정 그룹으로 나눌 때 필요한 함수입니다.
만약에 `SELECT` 절에서 몇 개의 컬럼을 조회 할 경우  SQL에서는 직,간접적으로 이 컬럼들이  기본키(primary key)를 기준으로 
`GROUP BY`에 종속되도록 합니다.
- `HAVING` 는 `GROUP BY`로 만들어진 그룹들에 대해 조건을 적용 할 경우 필요합니다.(그룹에 대해 조건을 걸기 때문에 `WHERE`절과 비슷합니다)

 SQLAlchemy에서는 ` Select.group_by()`와 `Select.having()`를 이용해 `GROUP BY`와 `HAVING`을 구현 할 수 있습니다.

```python
>>> with engine.connect() as conn:
...     result = conn.execute(
...         select(User.name, func.count(Address.id).label("count")).
...         join(Address).
...         group_by(User.name).
...         having(func.count(Address.id) > 1)
...     )
...     print(result.all())

> 

Reviewer 2: ```suggestion
""" 위 구문은 아래 SQL을 표현합니다.
SELECT user_account.name, count(address.id) AS count
FROM user_account JOIN address ON user_account.id = address.user_id GROUP BY user_account.name
HAVING count(address.id) > ?
[...] (1,)
"""

```

```python
>>> from sqlalchemy import func, desc
>>> # 컬럼에도 num_addresses 별칭이 들어가고 order_by에도 같은 별칭이 들어갑니다.
>>> stmt = select(
...         Address.user_id,
...         func.count(Address.id).label('num_addresses')).\
...         group_by("user_id").order_by("user_id", desc("num_addresses"))
>>> print(stmt)
"""
SELECT address.user_id, count(address.id) AS num_addresses
FROM address GROUP BY address.user_id ORDER BY address.user_id, num_addresses DESC
"""
```
<br>

## 별칭 사용하기

여러개의 테이블을 `JOIN`을 이용해 조회 할 경우 쿼리문에서 테이블 이름을 여러번 적어줘야 하는 경우가 많습니다.
SQL에서는 이러한 문제를 테이블 명이나 서브 쿼리에 별칭(aliases)를 지어 반복되는 부분을 줄일 수 있습니다.

한편 SQLAlchemy에서는 이러한 별칭들은 Core의 ` FromCaluse.alias()`함수를 이용해 구현 할 수 있습니다.
`Table`객체 네임스페이스 안에 `Column`객체가 있어 `Table.c`로 컬럼명에 접근할 수 있었는데요.
```python
print(select(user_table.c.name, user_table.c.fullname))
"""
SELECT user_account.name, user_account.fullname
FROM user_account
"""
``` 
`Alias`객체 네임스페이스 안에도 `Column`객체가 있어 `Alias.c`로 컬럼에 접근 가능합니다.
```python
>>> # user_alias_1과 user_alias_2 모두 Alias객체입니다.
>>> user_alias_1 = user_table.alias(‘table1’) 
>>> user_alias_2 = user_table.alias(‘table2’)

> >>> # 방금 만든 테이블 별명으로 컬럼에 접근하려면 Alia.c.컬럼명으로 접근해야합니다.

Reviewer 2: ```suggestion
>>> # 방금 만든 테이블 별명으로 컬럼에 접근하려면 Alias.c.컬럼명으로 접근해야합니다.
```

>>> print(
...     select(user_alias_1.c.name, user_alias_2.c.name).
...     join_from(user_alias_1, user_alias_2, user_alias_1.c.id > user_alias_2.c.id)
... )

"""
SELECT table1.name, table2.name AS name_1 
FROM user_account AS table1 JOIN user_account AS table2 ON table1.id > table2.id
"""
```

<br>

### ORM 엔티티 별칭

 ORM도  `FromClause.alias()`메서드와 비슷한 `aliased()`함수가 존재합니다.  

이 ORM `aliased()`는 ORM의 기능을 유지하면서 원래 매핑된 `Table`객체에 내부적으로 `Alias`객체를 생성합니다.

> 여기서 `Address`와 `User`객체는 [이전 장]()에서 만들어졌는데,
> 두 객체 모두  `Base` 클래스를  상속받아`Table`객체를 포함하고 있는 엔터티입니다.

```python
>>> user_alias_1 = user_table.alias()
>>> user_alias_2 = user_table.alias()
>>> # 예제에서는 User나 Address엔터티에 적용됩니다.  
>>> print(
...     select(User).
...     join_from(User, address_alias_1).
...     where(address_alias_1.email_address == 'patrick@aol.com').
...     join_from(User, address_alias_2).
...     where(address_alias_2.email_address == 'patrick@gmail.com')
... )
"""
SELECT user_account.id, user_account.name, user_account.fullname 
FROM user_account JOIN address AS address_1 ON user_account.id = address_1.user_id JOIN address AS address_2 ON user
"""
```

<br>

## 서브쿼리와 CTE

> 이 섹션에서는 일반적으로 SELECT의 FROM 절에 있는 서브 쿼리에 대해 설명합니다. 서브쿼리와 유사한 방식으로 사용되지만 추가 기능이 포함된 CTE도 다룹니다.

Reviewer 2: ```suggestion
이 섹션에서는 일반적으로 SELECT의 FROM 절에 있는 서브 쿼리에 대해 설명합니다. 서브쿼리와 유사한 방식으로 사용되지만 추가 기능이 포함된 CTE도 다룹니다.

> CTE(Common Table Expression)는 동일 쿼리 내에서 여러번 참조할 수 있게 하는 쿼리 내 임시 결과 집합입니다. [이 글](https://blog.sengwoolee.dev/84)에 CTE에 대해 잘 설명되어 있으니, 잘 모르시겠는 분들은 참고하시면 좋습니다.
```

SQLAlchemy는 `Subquery` 개체를 
`Select.subquery()`사용하여 서브 쿼리를 나타내고 `Select.cte()` 를 사용하여 CTE를 나타냅니다. 

```python

> >>> subq = select(
> ...     func.count(address_table.c.id).label("count"),
> ...     address_table.c.user_id
> ... ).group_by(address_table.c.user_id).subquery()
> 
> >>> stmt = select(
> ...    user_table.c.name,
> ...    user_table.c.fullname,
> ...    subq.c.count
> ... ).join_from(user_table, subq) # ON절은 두 개의 테이블이 이미 foreigh key로 제약이 걸려있어 자동 바인딩된다.
> 

Reviewer 2: ```suggestion
>>> subq = select(
...     func.count(address_table.c.id).label("count"),
...     address_table.c.user_id
... ).group_by(address_table.c.user_id).subquery()
>>> print(subq)
"""
SELECT count(address.id) AS count, address.user_id
FROM address GROUP BY address.user_id
"""

>>> stmt = select(
...    user_table.c.name,
...    user_table.c.fullname,
...    subq.c.count
... ).join_from(user_table, subq) # ON절은 두 개의 테이블이 이미 foreigh key로 제약이 걸려있어 자동 바인딩된다.
>>> print(stmt)
"""
SELECT user_account.name, user_account.fullname, anon_1.count
FROM user_account JOIN (SELECT count(address.id) AS count, address.user_id AS user_id
FROM address GROUP BY address.user_id) AS anon_1 ON user_account.id = anon_1.user_id
"""
```

```
<br>

### 계층 쿼리 

> SQLAlchemy에서 CTE 구문을 사용하는 방법은 서브 쿼리 구문이 사용되는 방식과 거의 동일합니다. 대신 `Select.subquery()` 메서드의 호출을 Select.cte()를 사용하도록 변경하여 결과 객체를 FROM 요소로 사용할 수 있습니다. 

Reviewer 2: ```suggestion
SQLAlchemy에서 CTE 구문을 사용하는 방법은 서브 쿼리 구문이 사용되는 방식과 거의 동일합니다. 대신 `Select.subquery()` 메서드의 호출을 `Select.cte()`를 사용하도록 변경하여 결과 객체를 FROM 요소로 사용할 수 있습니다. 
```

```python
>>> subq = select(
...     func.count(address_table.c.id).label("count"),
...     address_table.c.user_id
... ).group_by(address_table.c.user_id).cte()

>>> stmt = select(
...    user_table.c.name,
...    user_table.c.fullname,
...    subq.c.count

> ... ).join_from(user_table, subq)

Reviewer 2: ```suggestion
... ).join_from(user_table, subq)

>>> print(stmt)
"""
WITH anon_1 AS
(SELECT count(address.id) AS count, address.user_id AS user_id
FROM address GROUP BY address.user_id)
 SELECT user_account.name, user_account.fullname, anon_1.count
FROM user_account JOIN anon_1 ON user_account.id = anon_1.user_id
"""
```

```

<br>

### ORM 엔티티 서브쿼리, CTE
여기서는 `aliased()`가 `Subquery`, `CTE`서브 쿼리에 대해 동일한 작업을 수행하는 과정을 확인할 수 있습니다.

```python
>>> subq = select(Address).where(~Address.email_address.like('%@aol.com')).subquery()
>>> address_subq = aliased(Address, subq)
>>> stmt = select(User, address_subq).join_from(User, address_subq).order_by(User.id, address_subq.id)
>>> with Session(engine) as session:
...     for user, address in session.execute(stmt):

> ...         print(f"{user} {address}")

Reviewer 2: ```suggestion
...         print(f"{user} {address}")

""" 위 구문은 아래 쿼리를 표현합니다.
SELECT user_account.id, user_account.name, user_account.fullname,
anon_1.id AS id_1, anon_1.email_address, anon_1.user_id
FROM user_account JOIN
(SELECT address.id AS id, address.email_address AS email_address, address.user_id AS user_id
FROM address
WHERE address.email_address NOT LIKE ?) AS anon_1 ON user_account.id = anon_1.user_id
ORDER BY user_account.id, anon_1.id
[...] ('%@aol.com',)
"""

User(id=1, name='spongebob', fullname='Spongebob Squarepants') Address(id=1, email_address='spongebob@sqlalchemy.org')
User(id=2, name='sandy', fullname='Sandy Cheeks') Address(id=2, email_address='sandy@sqlalchemy.org')
User(id=2, name='sandy', fullname='Sandy Cheeks') Address(id=3, email_address='sandy@squirrelpower.org')
```

```
아래는 CTE생성자를 이용해 위와 같은  사용하는 예제입니다
```python
>>> cte = select(Address).where(~Address.email_address.like('%@aol.com')).cte()
>>> address_cte = aliased(Address, cte)
>>> stmt = select(User, address_cte).join_from(User, address_cte).order_by(User.id, address_cte.id)
>>> with Session(engine) as session:
...     for user, address in session.execute(stmt):
...         print(f"{user} {address}")

User(id=1, name='spongebob', fullname='Spongebob Squarepants') Address(id=1, email_address='spongebob@sqlalchemy.org')
User(id=2, name='sandy', fullname='Sandy Cheeks') Address(id=2, email_address='sandy@sqlalchemy.org')
User(id=2, name='sandy', fullname='Sandy Cheeks') Address(id=3, email_address='sandy@squirrelpower.org')
```
<br>

## 스칼라 서브 쿼리, 상호연관 쿼리 

스칼라 서브 쿼리에 대해 설명하기전 잠시 SQL에서 서브 쿼리에 대해 이야기 하겠습니다. [출처:바이헨 블로그](https://rinuas.tistory.com/entry/%EC%84%9C%EB%B8%8C%EC%BF%BC%EB%A6%ACSub-Query)

>  - 서브쿼리란 하나의 SQL문에 속한 `SELECT`문을 말하고 서브 쿼리의 바깥 쪽에 있는 SQL문을 `메인 쿼리`라고 합니다.

Reviewer 2: ```suggestion
 - "서브쿼리"란 하나의 SQL문에 속한 `SELECT`문을 말하고 서브 쿼리의 바깥 쪽에 있는 SQL문을 "메인 쿼리"라고 합니다.
```

이때 서브쿼리의 종류는 메인 쿼리 컬럼을 참조 여부, 서브쿼리의 선언 위치, 서브 쿼리 실행 결과 ROW수에 따라 종류가 나눠집니다

- 메인 쿼리 컬럼 참조 여부에 따른 구분
    - **상호 연관 서브쿼리** : 서브쿼리가 메인 쿼리 컬럼을 참조
    - 비상호 연관 서브 쿼리: 서브쿼리가 메인 쿼리의 컬럼을 참조하지 않고 독립적으로 수행하고 메인쿼리에 정보를 전달할 목적으로 사용됩니다.
- 서브쿼리 선언 위치에 따른 구분
    - **스칼라 서브 쿼리** : SELECT 문의 컬럼자리에 오는 서브 쿼리(상호 연관 서브쿼리)
    - 인라인 뷰 : FROM절 자리에 오는 서브 쿼리 (상호 연관 서브 쿼리)
    - 중첩 서브 쿼리 : Where절 자리에 오는 서브 쿼리 (비상호 연관 서브 쿼리)
- 서브 쿼리 실행 결과 ROW수에 따른 구분
  - 단일행 서브쿼리(서브 쿼리 연산결과 ROW1개) 
  - 단중행 서브쿼리(서브 쿼리 연산결과 ROW2개이상):IN, ANY, ALL, EXSITS 
  
<br>

SQLAlchemy에서 스칼라 서브 쿼리는 `ColumnElement`객체의 일부인 `ScalarSelect`를 사용하는 방면 일반 서브 쿼리는`FromClause`객체에 있는 `Subquery`를 사용합니다.

> 스칼라 서브쿼리는 앞에서 설명했던 [그룹 합수]()와 같이 쓰이고는 합니다.   

Reviewer 2: ```suggestion
스칼라 서브쿼리는 앞에서 설명했던 그룹 함수와 같이 쓰이고는 합니다.   
```

```python

> >>> # Select.scalar_subquery()를 이용해 구현한 스칼라 서브 쿼리

Reviewer 2: ```suggestion
# Select.scalar_subquery()를 이용해 구현한 스칼라 서브 쿼리
```

> >>> subq = select(func.count(address_table.c.id)).\
> ...             where(user_table.c.id == address_table.c.user_id).\
> ...             scalar_subquery()

Reviewer 2: ```suggestion
>>> subq = select(func.count(address_table.c.id)).
...             where(user_table.c.id == address_table.c.user_id).
...             scalar_subquery()
```

>>> print(subq) #ScalarSelect객체
"""
(SELECT count(address.id) AS count_1
FROM address, user_account
WHERE user_account.id = address.user_id)
"""
```
스칼라 서브 쿼리가 `user_account`와 `address`를 FROM절에서 렌더링하지만 
메인쿼리에 있는 `user_account`테이블이 있어서
스칼라 서브 쿼리에서는 `user_account` 테이블을 렌더링하지 않습니다.

```python
>>> stmt = select(user_table.c.name, subq.label("address_count"))
>>> print(stmt)
"""
SELECT user_account.name, (SELECT count(address.id) AS count_1
FROM address
WHERE user_account.id = address.user_id) AS address_count
FROM user_account
"""
```
한편 상호 연관 쿼리를 작성 할 때 테이블 간의 연결이 모호해질 수도 있습니다.
> 튜토리얼에 나와있는 상호 연관쿼리 예제는 제가 이해하지 못했습니다.
> 잘 아시는분은 이문서에 기여 부탁드립니다.
>

<br>

## UNION, UNION ALL 연산자들

SQL에서는 `UNION`, `UNION ALL`등으로 두 개의 `SELECT`문을 합치는 것을 의미합니다.  
아래와 같이 쿼리문을 실행 할 수도 있습니다. 
```sql
SELECT id FROM user_account
union 
SELECT email_address FROM address
```
그 외에도 집합 연산인 `INTERSECT`(교집합), `EXCEPT`(차집합)도  SQL에서 지원합니다.  

> SQLAlchemy에서 `Select`구문은 `union()`, `intersect()`, `except_()` 혹은 

Reviewer 2: ```suggestion
SQLAlchemy에서 `Select` 객체에 대하여 `union()`, `intersect()`, `except_()` 혹은 
```

이러한 함수들의 반환 값은 `CompoundSelect`인데 `Select`와 비슷하게 쓰일 수 있는 객체이지만 더 적은 메서드를 갖고 있습니다.  
`union_all()`의 반환값 `CompoundSelect`객체는 `Connection.execute()`로 실행될 수 있습니다.  

```python
>>> from sqlalchemy import union_all
>>> stmt1 = select(user_table).where(user_table.c.name == 'sandy')
>>> stmt2 = select(user_table).where(user_table.c.name == 'spongebob')
>>> u = union_all(stmt1, stmt2) #u는 CompoundSelect 객체입니다.
>>> with engine.connect() as conn:
...     result = conn.execute(u)
...     print(result.all())

[(2, 'sandy', 'Sandy Cheeks'), (1, 'spongebob', 'Spongebob Squarepants')]
```

`Subquery`객체를 만들기 위해 `Select`가 `SelectBase.subquery()`메서드를 제공하는 것처럼
`CompoundSelect`객체를 서브 쿼리로 비슷한 방식으로 사용 할 수 있습니다.
```python
>>> u_subq = u.subquery()
>>> stmt = (
...     select(u_subq.c.name, address_table.c.email_address).
...     join_from(address_table, u_subq).
...     order_by(u_subq.c.name, address_table.c.email_address)
... )
>>> with engine.connect() as conn:
...     result = conn.execute(stmt)
...     print(result.all())

[('sandy', 'sandy@sqlalchemy.org'), ('sandy','sandy@squirrelpower.org'), 
('spongebob', 'spongebob@sqlalchemy.org')]
```

<br>

## EXISTS 서브쿼리들

> SQLAlchemy는 `SelectBase.exists()`메서드를 통해 `Exists`객체를 만들어  EXISTS구문을 구현합니다.

Reviewer 2: ```suggestion
SQLAlchemy는 `SelectBase.exists()`메서드를 통해 `Exists`객체를 만들어 `EXISTS` 구문을 구현합니다.
```

```python
>>> # subq는 Exists객체입니다.
>>> subq = (
...     select(func.count(address_table.c.id)).
...     where(user_table.c.id == address_table.c.user_id).
...     group_by(address_table.c.user_id).
...     having(func.count(address_table.c.id) > 1)
... ).exists()
>>> print(subq)
"""
EXISTS (SELECT count(address.id) AS count_1 
FROM address, user_account 
WHERE user_account.id = address.user_id GROUP BY address.user_id 
HAVING count(address.id) > :count_2)
"""
>>> with engine.connect() as conn:
...     result = conn.execute(
...         select(user_table.c.name).where(subq)
...     )
...     print(result.all())

[('sandy',)]
```

> 한편 EXISTS 구문은 부정으로 사용되지 않는 경우가 더 많습니다.

Reviewer 2: ```suggestion
한편 `EXISTS` 구문은 부정으로 사용되지 않는 경우가 더 많습니다.
```

```python
# 이메일 주소가 없는 유저네임을 선택하는 쿼리문입니다.
# "~" 연산이 들어간 부분을 확인해보세요
>>> subq = (
...     select(address_table.c.id).
...     where(user_table.c.id == address_table.c.user_id)
... ).exists()
>>> stmt = select(user_table.c.name).where(~subq)
>>> print(stmt)
"""
SELECT user_account.id 
FROM user_account 
WHERE NOT (EXISTS (SELECT count(address.id) AS count_1 
FROM address 
WHERE user_account.id = address.user_id GROUP BY address.user_id 
HAVING count(address.id) > :count_2))
"""
>>> with engine.connect() as conn:

> ...     result = conn.execute(
> ...        stmt 
> ...     )
> ...     print(result.all())

Reviewer 2: ```suggestion
...     result = conn.execute(stmt)
...     print(result.all())

[('patrick',)]
```

```

<br>

## SQL 함수 다뤄보기

이 섹션 앞부분의 [그룹함수 :GROUP BY, HAVING]()에서 처음 소개된 `func` 객체는
새로운 `Function` 객체를 생성하기 위한 팩토리 역할을 합니다.  
 `select()`와 같은 구문을 사용 할때는 인자 값으로 `func`객체로 생성된 SQL함수를 받을 수 있습니다.

- `count()` : 집계함수로 행의 개수를 출력하는데 사용됩니다.
  ```python
  >>> # cnt는 <class 'sqlalchemy.sql.functions.count'>타입입니다.
  >>> cnt = func.count() 
  >>> print(select(cnt).select_from(user_table))
  """
  SELECT count(*) AS count_1FROM user_account
  """
  ```
- `lower()` : 문자열 함수로 문자열을 소문자로 바꿔줍니다.
  ```python
  >>> print(select(func.lower("A String With Much UPPERCASE")))
  """
  SELECT lower(:lower_2) AS lower_1
  """
  ```
- `now()` : 현재 시간과 날짜를 반환해주는 함수입니다. 
  이 함수는 굉장히 흔하게 사용되는 함수이기에 SQLAlchemy는 서로 다른 백엔드에서 손쉽게 렌더링 할 수 있도록 도와줍니다. 
  ```` python

>   #SQLite에서 CURRENT_TIMESTAMP함수로 쓰일 경우 

Reviewer 2: ```suggestion
```

  >>> stmt = select(func.now())
  >>> with engine.connect() as conn:
  ...     result = conn.execute(stmt)

>   ...     print(result.all())

Reviewer 2: ```suggestion
  ...     print(result.all())

  [(datetime.datetime(...),)]
```

> 이에 대해 이해하신 분 있으시다면 이 부분에 기여 부탁드립니다. 감사합니다.

<br> 

### Built-in Functions Have Pre-Configured Return Types
> 원문의 Built-in Functions Have Pre-Configured Return Types 부분은 제가 이해하지 못했습니다. 
> 이에 대해 이해하신 분 있으시다면 이 부분에 기여 부탁드립니다. 감사합니다.

<br> 

### 윈도우 함수

윈도우 함수는 GROUP BY와 비슷한 함수이고 행간의 관계를 쉽게 정의 하기 위해 만든 함수입니다.  
윈도우 함수에 대해 알고 싶으신 분들은 [민지님 블로그](https://mizykk.tistory.com/121)에 자세한 설명이 나와있으니 한 번 읽어보시고 아래의 내용들을 이어서 읽어주세요.  

SQLAlchemy에서는, `func` 네임스페이스에 의해 생성된 모든 SQL 함수  중 하나로
OVER 구문을 구현하는  `FunctionElement.over()` 메서드가 있습니다.

윈도우 함수 중 하나로 행의 개수를 세는 `row_number()` 함수가 있습니다.  
각 행을 사용자 이름대로 그룹을 나누고 그 안에서 이메일 주소에 번호를 매길 수 있습니다.

```python
# FunctionElement.over.partition_by파라미터를 사용하여 
# PARTITION BY 절이 OVER 절에 렌더링되도록 했습니다.
>>> stmt = select(
...     func.row_number().over(partition_by=user_table.c.name),
...     user_table.c.name,
...     address_table.c.email_address
... ).select_from(user_table).join(address_table)
>>> with engine.connect() as conn:  
...     result = conn.execute(stmt)
...     print(result.all())
[(1, 'sandy', 'sandy@sqlalchemy.org'), 
 (2, 'sandy', 'sandy@squirrelpower.org'), 
 (1, 'spongebob', 'spongebob@sqlalchemy.org')]
```

> `FunctionElement.over.order_by`를 사용하여 ORDER BY 절을 사용할 수도 있습니다.

Reviewer 2: ```suggestion
`FunctionElement.over.order_by`를 사용하여 `ORDER BY` 절을 사용할 수도 있습니다.
```

```python
>>> stmt = select(
...     func.count().over(order_by=user_table.c.name),
...     user_table.c.name,
...     address_table.c.email_address).select_from(user_table).join(address_table)
>>> with engine.connect() as conn:  
...     result = conn.execute(stmt)
...     print(result.all())

[(2, 'sandy', 'sandy@sqlalchemy.org'), 
 (2, 'sandy', 'sandy@squirrelpower.org'), 
 (3, 'spongebob', 'spongebob@sqlalchemy.org')]
```
<br> 

### WITHIN GROUP, FILTER등 특수한 지정자

`WITHIN GORUP`이라는 SQL 구문은 순서 집합 또는 가상 집합 그리고 집계함수와 함께 쓰입니다.
일반적인 순서 집합 함수는 `percentile_cont()` 그리고 `rank()`를 포함하고 있습니다.
SQLAlchemy에서는 `rank`, `dense_rank`, `percentile_count`, `percentile_disc`가 구현되어 있고

> 각 각은 `FunctionElement.within_group()`메서드를 갖고 있습니다.

Reviewer 2: ```suggestion
각각은 `FunctionElement.within_group()`메서드를 갖고 있습니다.
```

```
그래서 보통은 `text()`보다 `literal_column()`을 사용해 작은 따옴표가 결과물에 붙어져서 나오는 문제를 해결할 수 있습니다.  
여기에서  `text`와 `literal_column()`은 거의 비슷하지만  `literal_column()`은 명시적으로 컬럼을 의미하고, 서브쿼리나 다른 SQL 표현식에서 쓰일 수 있게 라벨링도 할 수 있습니다.  

```python
>>> from sqlalchemy import literal_column
>>> stmt = (
...     select(
...         literal_column("'some phrase'").label("p"), user_table.c.name
...     ).order_by(user_table.c.name)
... )
>>> with engine.connect() as conn:
...     for row in conn.execute(stmt):
...         print(f"{row.p}, {row.name}")
some phrase, patrick
some phrase, sandy
some phrase, spongebob

```

<br>

## WHERE절

SQLAlchemy를 사용하면 Python 연산자를 사용하여 `name` = 'thead' 또는 `user_id` > 10인 데이터만 출력하는 쿼리를 쉽게 작성할 수 있습니다.

```python
>>> print(user_table.c.name == 'squidward')
user_account.name = :name_1

>>> print(address_table.c.user_id > 10)
address.user_id > :user_id_1
```

`WHERE`절을 만들기 위해 `Select.where()`메서드안에 인자값을 넣어 사용할 수도 있습니다.

```python
>>> print(select(user_table).where(user_table.c.name == 'squidward'))
"""
SELECT user_account.id, user_account.name, user_account.fullname
FROM user_account
WHERE user_account.name = :name_1
"""
```

where 절을 통한 join을 구현 할 경우 아래와 같이 작성 가능합니다.
```python

> >>> select(address_table.c.email_address)

Reviewer: ```suggestion
>>> select(address_table.c.email_address).
```
.이 안 붙은거 같아서 작게나마 적어봅니다.
